bcm283x/gpio.go

```diff
@@ -1020,6 +1024,64 @@ var mapping = [][6]pin.Func{
 	{""},
 }
 
+// BCM2711 specific alternate function mapping; mostly compatible to
+// mapping283x, with the exception that there is no user-accessible SPI2
+// (the original SPI2-related positions have been assigned
+// corresponding SPI0 functions). Also, since there is an additional
+// PWM1 device, the original PWM0/PWM1 entries had to be
```

Comment:
I'd prefer to keep PWM0 / PWM1. Take for example UART0_TX, it's present multiple times. The important point if it is the same signal or not.

Reply:
Just to be sure: Previously, in the BCM2835 mapping there only was _one_ PWM peripheral with two channels, PWM0 and PWM1 (each occuring as functions in multiple pins) -- PWM0 and PWM1 are the names of the channels.

The BCM2711, on the other hand, contains _two_ PWM peripherals, each with two channels. Here PWM0 and PWM1 are the names of the peripherals (PWM blocks), each containing two channels.  
So there are _four_ PWM channels in case of the BCM2711, as compared to _two_ on the BCM2835.

The mapping of the PWM channels to GPIOs' alternate functions is nearly identical between BCM2835 and BCM2711, with the exception that in case of the latter at GPIO40/41 the second PWM peripheral, PWM1, is exposed instead of the first PWM peripheral. 

To be sure I'm understanding your intention regarding the renaming right: should only PWM0_0 and PWM0_1 be renamed to PWM0 and PWM1, and PWM1_0 and PWM1_1  (at GPIO40/41) **kept this way,**
or shall also the `1_` of the second PWM peripheral be omitted?

According to the first understanding, after the renaming the table would contain occurences of `PWM0`, `PWM1`, `PWM1_0`, and `PWM1_1`.  
According to the second understanding, the table would only contain occurences of  `PWM0`, `PWM1`.

Which of the two did you have in mind?

Reply:
ohh that's cool. It's fine then but I fear there's a hack or two in the code that assumes PWMn.
I don't have a strong opinion. (Still thinking)

Reply:
> After a quick check:
> 
> * Update `aliases` to match if we use PWM0_0.

I found the following to work with both bcm283x and bcm2711, with regard to gpioreg.Aliases():

```diff
@@ -1435,18 +1438,26 @@ func (d *driverGPIO) Init() (bool, error) {
 	// Register some BCM-documentation specific names.
 	// Do not do UARTx_TXD/RXD nor the PCM_xxx ones.
 	aliases := [][2]string{
 		{"GPCLK0", "CLK0"},
 		{"GPCLK1", "CLK1"},
 		{"GPCLK2", "CLK2"},
 		{"PWM0_OUT", "PWM0"},
 		{"PWM1_OUT", "PWM1"},
 	}
+	if isBCM2711 {
+		aliases = append(aliases[:3], [][2]string{
+			{"PWM0_OUT", "PWM0_0"},
+			{"PWM1_OUT", "PWM0_1"},
+			{"PWM0", "PWM0_0"},
+			{"PWM1", "PWM0_1"},
+		}...)
+	}
```
I thought about whether for the nonce it may be preferable to keep `PWM0` and `PWM1` alternative function names regardless of whether  it refers to bcm2711's first or second PWM peripheral. Then `mapping2711` would show `PWM0` and `PWM1` at the same positions as for bcm283x, and programs using it would work the same as before.
There is one more thing I noticed: The current pwm dma code likely won't work for GPIO40/41 of bcm2711, because this code appears to be written with _one_ pwm peripheral in mind. For GPIO40/41 on the bcm2711, the second peripheral, PWM1, the DREQ 1 (DSI/PWM1) would have to be configured in the DMA controller instead of 5 (PWM0).

With this in mind, I would suggest dropping the peripheral index `0_` or `1_` in names like PWM0_1 in the new `mapping2711` for the time being and continue to use PWM0, PWM1 as you preferred in your first comment, and also keep aliases unchanged (until, if needed, a future change will address GPIO40/41 PWM functionality on the BCM2711).
From my view, the impact of this patch would be smaller, and compatibility would be maintained better, than in case of introducing `PWM0_*` and `PWM1_*` and modifying `aliases`.

So, if you agree I would just convert the PWM alt funcs to PWM0 and PWM1 according to your first comment.
(Otherwise I would continue with the `aliases` approach).